### ISSUE TITLE
socket.error: (19, 'No such device')

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10920]
|[<img alt="ghazel's avatar" src="https://avatars.githubusercontent.com/u/31402?s=50" width="50" height="50">](https://github.com/ghazel)| @ghazel reported|
|-|-|
|Trac ID|trac#1345|
|Type|defect|
|Created|2005-12-01 00:14:34Z|

<details><summary>Searchable metadata</summary>

```
trac-id__1345 1345
type__defect defect
reporter__ghazel ghazel
priority__high high
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__win32 win32
time__1133396074000000 1133396074000000
changetime__1264692059000000 1264692059000000
version__ 
owner__ 
cc__justinj cc__ghazel cc__teratorn
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="ghazel's avatar" src="https://avatars.githubusercontent.com/u/31402?s=50" width="50" height="50">](https://github.com/ghazel)|@ghazel commented|
|-|-|

```
#!html
<pre>
here's the traceback and the log.

Unhandled error in Deferred:
(debug:  C: Deferred was created:
 C:  File "twisted_traceback.py", line 7, in ?
 C:    p.joinGroup('239.255.255.250', socket.INADDR_ANY)
 C:  File "/usr/lib/python2.4/site-packages/twisted/internet/udp.py", line 331, 
in joinGroup
 C:    return self.reactor.resolve(addr).addCallback(self._joinAddr1, 
interface, 1)
 C:  File "/usr/lib/python2.4/site-packages/twisted/internet/base.py", line 
309, in resolve
 C:    return defer.succeed(name)
 C:  File "/usr/lib/python2.4/site-packages/twisted/internet/defer.py", line 
52, in succeed
 C:    d = Deferred()
 I: First Invoker was:
 I:  File "twisted_traceback.py", line 7, in ?
 I:    p.joinGroup('239.255.255.250', socket.INADDR_ANY)
 I:  File "/usr/lib/python2.4/site-packages/twisted/internet/udp.py", line 331, 
in joinGroup
 I:    return self.reactor.resolve(addr).addCallback(self._joinAddr1, 
interface, 1)
 I:  File "/usr/lib/python2.4/site-packages/twisted/internet/base.py", line 
309, in resolve
 I:    return defer.succeed(name)
 I:  File "/usr/lib/python2.4/site-packages/twisted/internet/defer.py", line 
53, in succeed
 I:    d.callback(result)
)
Traceback (most recent call last):
  File "/usr/lib/python2.4/site-packages/twisted/internet/defer.py", line 328, 
in _runCallbacks
    self.result = callback(self.result, *args, **kw)
  File "/usr/lib/python2.4/site-packages/twisted/internet/udp.py", line 334, in 
_joinAddr1
    return self.reactor.resolve(interface).addCallback(self._joinAddr2, addr, 
join)
  File "/usr/lib/python2.4/site-packages/twisted/internet/defer.py", line 216, 
in addCallback
    callbackKeywords=kw)
  File "/usr/lib/python2.4/site-packages/twisted/internet/defer.py", line 207, 
in addCallbacks
    self._runCallbacks()
--- &lt;exception caught here> ---
  File "/usr/lib/python2.4/site-packages/twisted/internet/defer.py", line 328, 
in _runCallbacks
    self.result = callback(self.result, *args, **kw)
  File "/usr/lib/python2.4/site-packages/twisted/internet/udp.py", line 343, in 
_joinAddr2
    self.socket.setsockopt(socket.IPPROTO_IP, cmd, addr + interface)
  File "&lt;string>", line 1, in setsockopt
    
socket.error: (19, 'No such device')


Here's the code:
import socket
from twisted.internet import reactor
from twisted.internet.protocol import DatagramProtocol
import twisted.internet
twisted.internet.defer.setDebugging(True)
p = reactor.listenMulticast(8000, DatagramProtocol(), '127.0.0.1')
p.joinGroup('239.255.255.250', socket.INADDR_ANY)
reactor.run()
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="teratorn's avatar" src="https://avatars.githubusercontent.com/u/154069?s=50" width="50" height="50">](https://github.com/teratorn)|@teratorn commented|
|-|-|

This code seems to run fine for me (WinXP SP2). What Twisted version? What Windows version? Have you tried the latest Twisted?

**Comment by twisted-trac** [comment]
|[<img alt="ghazel's avatar" src="https://avatars.githubusercontent.com/u/31402?s=50" width="50" height="50">](https://github.com/ghazel)|@ghazel commented|
|-|-|

I assume it only happens if you don't have an internet connection. I've submitted a patch elsewhere which wraps the self.socket.setsockopt call in a try. except socket.error block that catches this and other errors and returns 0 instead of crashing.

Perhaps that has already been accepted.

**Comment by twisted-trac** [comment]
|[<img alt="teratorn's avatar" src="https://avatars.githubusercontent.com/u/154069?s=50" width="50" height="50">](https://github.com/teratorn)|@teratorn commented|
|-|-|

justinj, are you still looking in to this? These days udp.py will catch this error and raise a MulticastJoinError that wraps the underlying Exception.

Can anyone confirm that this only happens if you don't have an internet connection? If so then it seems like a perfectly reasonable thing for the joinGroup() deferred to errback, and we should close this ticket.

**Comment by twisted-trac** [comment]
|<img alt="justinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|justinj commented|
|-|-|

No, I haven't had any time to work on Twisted for a while now.  The only Windows box I have access to is at work.  At this point it is safe to assume all of my tickets should be reassigned to someone else.

**Comment by twisted-trac** [comment]
|<img alt="justinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|justinj removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `closed`|
|-|-|

This was fixed in r16454, for [#11108](https://github.com/twisted/twisted/issues/11108), which was more or less a duplicate of this.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
